scylla-cql/src/types/serialize/row.rs

```diff
@@ -892,6 +901,123 @@ mod tests {
+        match err.0.downcast_ref() {
+            Some(err) => err,
+            None => panic!("not a BuiltinTypeCheckError: {}", err),
+        }
+    }
+
+    fn get_ser_err(err: &SerializationError) -> &BuiltinSerializationError {
+        match err.0.downcast_ref() {
+            Some(err) => err,
+            None => panic!("not a BuiltinSerializationError: {}", err),
+        }
+    }
+
+    #[test]
+    fn test_tuple_errors() {
+        // Unit
+        #[allow(clippy::let_unit_value)] // The let binding below is intentional
+        let v = ();
+        let spec = [col("a", ColumnType::Text)];
+        let err = do_serialize_err(v, &spec);
+        let err = get_typeck_err(&err);
+        assert_eq!(err.rust_name, std::any::type_name::<()>());
+        assert!(matches!(
+            err.kind,
+            BuiltinTypeCheckErrorKind::WrongColumnCount {
+                actual: 0,
+                asked_for: 1,
+            }
+        ));
+
+        // Non-unit tuple
+        // Count mismatch
+        let v = ("Ala ma kota",);
+        let spec = [col("a", ColumnType::Text), col("b", ColumnType::Text)];
+        let err = do_serialize_err(v, &spec);
+        let err = get_typeck_err(&err);
+        assert_eq!(err.rust_name, std::any::type_name::<(&str,)>());
+        assert!(matches!(
+            err.kind,
+            BuiltinTypeCheckErrorKind::WrongColumnCount {
+                actual: 1,
+                asked_for: 2,
+            }
+        ));
+
+        // Serialization of one of the element fails
+        let v = ("Ala ma kota", 123_i32);
+        let spec = [col("a", ColumnType::Text), col("b", ColumnType::Text)];
+        let err = do_serialize_err(v, &spec);
+        let err = get_ser_err(&err);
+        assert_eq!(err.rust_name, std::any::type_name::<(&str, i32)>());
+        let BuiltinSerializationErrorKind::ColumnSerializationFailed { name, err: _ } = &err.kind;
+        assert_eq!(name, "b");
+    }
+
+    #[test]
+    fn test_slice_errors() {
+        // Non-unit tuple
+        // Count mismatch
+        let v = vec!["Ala ma kota"];
+        let spec = [col("a", ColumnType::Text), col("b", ColumnType::Text)];
+        let err = do_serialize_err(v, &spec);
+        let err = get_typeck_err(&err);
+        assert_eq!(err.rust_name, std::any::type_name::<Vec<&str>>());
+        assert!(matches!(
+            err.kind,
+            BuiltinTypeCheckErrorKind::WrongColumnCount {
+                actual: 1,
+                asked_for: 2,
+            }
+        ));
+
+        // Serialization of one of the element fails
+        let v = vec!["Ala ma kota", "Kot ma pchły"];
+        let spec = [col("a", ColumnType::Text), col("b", ColumnType::Int)];
+        let err = do_serialize_err(v, &spec);
+        let err = get_ser_err(&err);
+        assert_eq!(err.rust_name, std::any::type_name::<Vec<&str>>());
+        let BuiltinSerializationErrorKind::ColumnSerializationFailed { name, err: _ } = &err.kind;
```

Comment:
I didn't know this code is possible - what happends if the variant is different than `ColumnSerializationFailed`? `BuiltinSerializationErrorKind` is `#[non_exhaustive]` so adding a new variant is backwards compatible afaiu

Reply:
`ColumnSerializationFailed` is the only variant of `BuiltinSerializationErrorKind`, so this pattern is irrefutable and there is no need for `else`. In fact, I tried putting an `else {}` to future-proof the expression, and while `rustc` ate it, `clippy` complained that it is not needed.

As far as `#[non_exhaustive]` goes, it only affects other crates than the current one - which makes sense; if you add an enum variant, you can immediately fix all the uses of the enum in the same crate.

Reply:
Wow, I didn't expect Rust to be that smart in this regard, nice.